### DISCUSSION TITLE
Block reference to inserted block

### DISCUSSION THREAD

**Discussion by Aaron12300** [haplokuon/netDxf#340] [Q&A] [answered]
Wonderful code, truly.. Glad I found this. Question please.
I want to have a dwg for someone that has several blocks representing some "things". Named differently like Thing A, Thing B ect.
I would pick the "thing" that I want on that design and place it properly. Export to DXF and parse what thing it is and where its located, maybe a user attribute or 2 as well. This code got me 99% of the way there, but I cannot easily figure out what blocks are my "Thing A" block because it seems AutoCad renames the block to an *U name. I have found them (I inserted 2 of the same "Thing") and used wierd coordinates to find them in the INSERT section, but there is no reference back to the original BLOCK. I think AutoCad just uses the block to create a new one and thats it.. BUT, if I show properties in AutoCad, the block(s) is/are called the original block name.
Is there a ref back to the original block definition? End purpose is to send to a machine that knows what a Thing A is and needs location and some other stuff.. I know this is not exactly a question about your software, but after reading your code and using it, I think you qualify as the worlds foremost expert in DXF format... :) Thanks!!!

**Comment by haplokuon** [comment, marked as answer]
From what you wrote in your post you are working with dynamic blocks, and what you describe this is its normal behavior. Although netDxf can import them you will loose all its dynamic parameters, but you can still retrieve the information you seek. It is stored in the extended data information of the block record under an application registry called "AcDbBlockRepBTag", you can do something like this:
```c#
// netDxf can read dynamic blocks but you will loose the information related with the parameters, actions and constrains
// for every insert AutoCad creates a new block definition with the name *U#, where # is a positive integer
// except in the case where the dynamic block parameters are not modified, in this case the original block will be used instead
DxfDocument drawing = DxfDocument.Load("YourDrawing.dxf");

// a dynamic block reference
Insert insert = drawing.Entities.Inserts.ElementAt(0);
// block referenced by the Insert
Block block = insert.Block;

// to access the original dynamic block we need to get first the extended data associated with the BlockRecord,
// the application registry for this extended data always has the name "AcDbBlockRepBTag"
// only the modified dynamic blocks "U#" store this handle, the original block stores other things of unknown use
XData xdata = block.Record.XData["AcDbBlockRepBTag"];
string handle = string.Empty;
// the original dynamic block record handle is stored in the extended data,
// use with care this is undocumented stuff, it appears to only store this value.
foreach (XDataRecord data in xdata.XDataRecord)
{
	if (data.Code == XDataCode.DatabaseHandle)
	{
		handle = (string) data.Value;
	}
}

// now we can get the original dynamic block record
BlockRecord originalDynamicBlockRecord = (BlockRecord) drawing.GetObjectByHandle(handle);
// this is the name of the original block
string dynamicBlockName = originalDynamicBlockRecord.Name;
// if we need the original block, we can get it from the list of blocks since we know its name
Block originalBlock = drawing.Blocks[dynamicBlockName];
```
You can also make a search for "dynamic blocks" in this repository, there are other post talking about them.

One additional note: the blocks generated by an array also use the same "U#" nomenclature but only the ones belonging to dynamic blocks will store that extended data information.
